In [133]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [134]:
url1 = "datasets/books.csv"
url2 = "datasets/ratings.csv"
url3 = "datasets/users.csv"

books = pd.read_csv(url1, encoding='latin-1')
ratings = pd.read_csv(url2, encoding='latin-1')
users = pd.read_csv(url3, encoding='latin-1')
books.head()

C:\Users\sadma\AppData\Local\Temp\ipykernel_12740\2952862037.py:5: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv(url1, encoding='latin-1')


ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [135]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [136]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [137]:
books.columns = books.columns.str.lower().str.replace(" ", "_").str.replace("-", "_")
ratings.columns = ratings.columns.str.lower().str.replace(" ", "_").str.replace("-", "_")
users.columns = users.columns.str.lower().str.replace(" ", "_").str.replace("-", "_")

In [138]:
books.columns

Index(['isbn', 'book_title', 'book_author', 'year_of_publication', 'publisher',
       'image_url_s', 'image_url_m', 'image_url_l'],
      dtype='object')

In [139]:
ratings.columns


Index(['user_id', 'isbn', 'book_rating'], dtype='object')

In [140]:
users.columns

Index(['user_id', 'location', 'age'], dtype='object')

In [141]:
books.rename(columns={
    "book_title": "book_name",
    "book_author": "author",
    "image_url_l": "image_url"
    }, inplace=True)
books.columns

Index(['isbn', 'book_name', 'author', 'year_of_publication', 'publisher',
       'image_url_s', 'image_url_m', 'image_url'],
      dtype='object')

In [142]:
books.drop([
    "image_url_s",
    "image_url_m",
    "year_of_publication"
    ], axis=1, inplace=True)
books.columns

Index(['isbn', 'book_name', 'author', 'publisher', 'image_url'], dtype='object')

In [143]:
ratings.rename(columns={
    "book_rating": "ratings"
    }, inplace=True)
ratings.columns


Index(['user_id', 'isbn', 'ratings'], dtype='object')

In [144]:
users.drop(columns={
    "age"
    }, inplace=True)
users.columns

Index(['user_id', 'location'], dtype='object')

In [145]:
books.head()

isbn                                          book_name  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author                   publisher  \
0    Mark P. O. Morford     Oxford University Press   
1  Richard Bruce Wright       HarperFlamingo Canada   
2          Carlo D'Este             HarperPerennial   
3      Gina Bari Kolata        Farrar Straus Giroux   
4       E. J. W. Barber  W. W. Norton &amp; Company   

                                           image_url  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [146]:
ratings.head()

user_id        isbn  ratings
0   276725  034545104X        0
1   276726  0155061224        5
2   276727  0446520802        0
3   276729  052165615X        3
4   276729  0521795028        6

In [147]:
users.head()

user_id                            location
0        1                  nyc, new york, usa
1        2           stockton, california, usa
2        3     moscow, yukon territory, russia
3        4           porto, v.n.gaia, portugal
4        5  farnborough, hants, united kingdom

In [148]:
print(books.shape)
print(ratings.shape)
print(users.shape)

(271360, 5)
(1149780, 3)
(278858, 2)


#### There are 271360 books 
#### There are 278858 users
### 278858 users rated 1149780

In [149]:
books.isnull().sum()

isbn         0
book_name    0
author       2
publisher    2
image_url    3
dtype: int64

In [150]:
books.dropna(inplace=True)

In [151]:
books.isnull().sum()

isbn         0
book_name    0
author       0
publisher    0
image_url    0
dtype: int64

In [152]:
ratings.isnull().sum()

user_id    0
isbn       0
ratings    0
dtype: int64

In [153]:
users.isnull().sum()

user_id     0
location    0
dtype: int64

In [154]:
books.duplicated().sum()

np.int64(0)

In [155]:
ratings.duplicated().sum()

np.int64(0)

In [156]:
users.duplicated().sum()

np.int64(0)

In [157]:
rating_w_books = ratings.merge(books, on="isbn")
rating_w_books.head()

user_id        isbn  ratings  \
0   276725  034545104X        0   
1   276726  0155061224        5   
2   276727  0446520802        0   
3   276729  052165615X        3   
4   276729  0521795028        6   

                                           book_name           author  \
0                               Flesh Tones: A Novel       M. J. Rose   
1                                   Rites of Passage       Judith Rae   
2                                       The Notebook  Nicholas Sparks   
3                                     Help!: Level 1    Philip Prowse   
4  The Amsterdam Connection : Level 4 (Cambridge ...      Sue Leather   

                    publisher  \
0            Ballantine Books   
1                      Heinle   
2                Warner Books   
3  Cambridge University Press   
4  Cambridge University Press   

                                           image_url  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/0155061224.0...  
2  http://images.amazon.com/images/P/0446520802.0...  
3  http://images.amazon.com/images/P/052165615X.0...  
4  http://images.amazon.com/images/P/0521795028.0...

In [158]:
grouped_books = rating_w_books.groupby("book_name")
grouped_books

In [159]:
total_ratings = grouped_books["ratings"].count()
total_ratings = total_ratings.reset_index()
total_ratings.rename({
    "ratings": "total_ratings"
    }, axis=1, inplace=True)
total_ratings.head()

book_name  total_ratings
0   A Light in the Storm: The Civil War Diary of ...              4
1                              Always Have Popsicles              1
2               Apple Magic (The Collector's series)              1
3   Ask Lily (Young Women of Faith: Lily Series, ...              1
4   Beyond IBM: Leadership Marketing and Finance ...              1

In [160]:
avg_ratings = grouped_books["ratings"].mean()
avg_ratings = avg_ratings.reset_index()
avg_ratings.rename({
    "ratings": "avg_ratings"
    }, axis=1, inplace=True)
avg_ratings.head()

book_name  avg_ratings
0   A Light in the Storm: The Civil War Diary of ...         2.25
1                              Always Have Popsicles         0.00
2               Apple Magic (The Collector's series)         0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...         8.00
4   Beyond IBM: Leadership Marketing and Finance ...         0.00

In [161]:
df = total_ratings.merge(avg_ratings, on="book_name")
df.head()

book_name  total_ratings  \
0   A Light in the Storm: The Civil War Diary of ...              4   
1                              Always Have Popsicles              1   
2               Apple Magic (The Collector's series)              1   
3   Ask Lily (Young Women of Faith: Lily Series, ...              1   
4   Beyond IBM: Leadership Marketing and Finance ...              1   

   avg_ratings  
0         2.25  
1         0.00  
2         0.00  
3         8.00  
4         0.00

In [162]:
df["total_ratings"] > 250

0         False
1         False
2         False
3         False
4         False
          ...  
241060    False
241061    False
241062    False
241063    False
241064    False
Name: total_ratings, Length: 241065, dtype: bool

In [163]:
df[df["total_ratings"] > 250]

book_name  total_ratings  \
764                                                  1984            284   
818                                   1st to Die: A Novel            509   
1048                                           2nd Chance            356   
1760                                   A Bend in the Road            346   
2281    A Child Called \It\": One Child's Courage to S...            265   
...                                                   ...            ...   
233844                           White Oleander : A Novel            387   
233845       White Oleander : A Novel (Oprah's Book Club)            356   
234734  Wicked: The Life and Times of the Wicked Witch...            326   
234945                                        Wild Animus           2502   
240854                                \O\" Is for Outlaw"            269   

        avg_ratings  
764        4.454225  
818        3.575639  
1048       3.269663  
1760       3.364162  
2281       4.086792  
...             ...  
233844     3.506460  
233845     3.772472  
234734     3.766871  
234945     1.019584  
240854     2.988848  

[186 rows x 3 columns]

In [164]:
df[df["total_ratings"] > 250].sort_values(by="avg_ratings", ascending=False)

book_name  total_ratings  \
80431   Harry Potter and the Prisoner of Azkaban (Book 3)            428   
80419        Harry Potter and the Goblet of Fire (Book 4)            387   
80438      Harry Potter and the Sorcerer's Stone (Book 1)            278   
80423   Harry Potter and the Order of the Phoenix (Boo...            347   
80411    Harry Potter and the Chamber of Secrets (Book 2)            556   
...                                                   ...            ...   
227695       Vinegar Hill (Oprah's Book Club (Paperback))            265   
233629                                           Whispers            286   
143373                                  Presumed Innocent            294   
94379                                        Isle of Dogs            288   
234945                                        Wild Animus           2502   

        avg_ratings  
80431      5.852804  
80419      5.824289  
80438      5.737410  
80423      5.501441  
80411      5.183453  
...             ...  
227695     2.245283  
233629     2.199301  
143373     2.139456  
94379      2.000000  
234945     1.019584  

[186 rows x 3 columns]

In [165]:
famous_books = df[df["total_ratings"] > 250].sort_values(by="avg_ratings", ascending=False).head(50)
famous_books

book_name  total_ratings  \
80431   Harry Potter and the Prisoner of Azkaban (Book 3)            428   
80419        Harry Potter and the Goblet of Fire (Book 4)            387   
80438      Harry Potter and the Sorcerer's Stone (Book 1)            278   
80423   Harry Potter and the Order of the Phoenix (Boo...            347   
80411    Harry Potter and the Chamber of Secrets (Book 2)            556   
191607  The Hobbit : The Enchanting Prelude to The Lor...            281   
187372  The Fellowship of the Ring (The Lord of the Ri...            368   
80442   Harry Potter and the Sorcerer's Stone (Harry P...            575   
211379     The Two Towers (The Lord of the Rings, Part 2)            260   
219736                              To Kill a Mockingbird            510   
183568                                  The Da Vinci Code            898   
187875                 The Five People You Meet in Heaven            430   
180552                             The Catcher in the Rye            449   
196321                          The Lovely Bones: A Novel           1295   
764                                                  1984            284   
144161                           Prodigal Summer: A Novel            253   
128667                                         Neverwhere            265   
206497                            The Secret Life of Bees            774   
168715  Stupid White Men ...and Other Sorry Excuses fo...            283   
223130  Tuesdays with Morrie: An Old Man, a Young Man,...            493   
204382                The Red Tent (Bestselling Backlist)            723   
191584               The Hitchhiker's Guide to the Galaxy            268   
129376   Nickel and Dimed: On (Not) Getting By in America            335   
93378                                       Into the Wild            252   
63864                                      Fahrenheit 451            409   
74747                           Girl with a Pearl Earring            526   
136142                                          Outlander            283   
233364  Where the Heart Is (Oprah's Book Club (Paperba...            585   
156098                     Seabiscuit: An American Legend            275   
107959                                         Life of Pi            664   
176841                                     The Bean Trees            389   
2281    A Child Called \It\": One Child's Courage to S...            265   
8433                                       ANGELA'S ASHES            279   
76340                                         Good in Bed            490   
64928   Fast Food Nation: The Dark Side of the All-Ame...            321   
12699                                       American Gods            302   
161641                                 Skipping Christmas            322   
105774  Left Behind: A Novel of the Earth's Last Days ...            318   
189546    The Golden Compass (His Dark Materials, Book 1)            336   
181675                                   The Color Purple            314   
160332                               Silence of the Lambs            256   
8751                                          About a Boy            262   
158134                  Seven Up (A Stephanie Plum Novel)            278   
175093  The Alchemist: A Fable About Following Your Dream            266   
80066   Hard Eight : A Stephanie Plum Novel (A Stephan...            269   
170097                       Suzanne's Diary for Nicholas            457   
111070                                  Lord of the Flies            259   
5664                              A Prayer for Owen Meany            413   
212065   The Vampire Lestat (Vampire Chronicles, Book II)            301   
233845       White Oleander : A Novel (Oprah's Book Club)            356   

        avg_ratings  
80431      5.852804  
80419      5.824289  
80438      5.737410  
80423      5.501441  
80411      5.183453  
191607     5.007117  
187372     4.9483

In [166]:
famous_books.merge(books, on="book_name")

book_name  total_ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)            428   
1    Harry Potter and the Prisoner of Azkaban (Book 3)            428   
2    Harry Potter and the Prisoner of Azkaban (Book 3)            428   
3         Harry Potter and the Goblet of Fire (Book 4)            387   
4         Harry Potter and the Goblet of Fire (Book 4)            387   
..                                                 ...            ...   
191                                  Lord of the Flies            259   
192                            A Prayer for Owen Meany            413   
193                            A Prayer for Owen Meany            413   
194   The Vampire Lestat (Vampire Chronicles, Book II)            301   
195       White Oleander : A Novel (Oprah's Book Club)            356   

     avg_ratings        isbn              author              publisher  \
0       5.852804  0439136350       J. K. Rowling             Scholastic   
1       5.852804  0439136369       J. K. Rowling             Scholastic   
2       5.852804  0786222743       J. K. Rowling        Thorndike Press   
3       5.824289  0439139597       J. K. Rowling             Scholastic   
4       5.824289  0439139600       J. K. Rowling  Scholastic Paperbacks   
..           ...         ...                 ...                    ...   
191     3.818533  0698102193  William G. Golding       Putnam Pub Group   
192     3.796610  0345361792         John Irving       Ballantine Books   
193     3.796610  0345363523         John Irving       Ballantine Books   
194     3.777409  0345313860           ANNE RICE       Ballantine Books   
195     3.772472  0316284955         Janet Fitch         Back Bay Books   

                                             image_url  
0    http://images.amazon.com/images/P/0439136350.0...  
1    http://images.amazon.com/images/P/0439136369.0...  
2    http://images.amazon.com/images/P/0786222743.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
4    http://images.amazon.com/images/P/0439139600.0...  
..                                                 ...  
191  http://images.amazon.com/images/P/0698102193.0...  
192  http://images.amazon.com/images/P/0345361792.0...  
193  http://images.amazon.com/images/P/0345363523.0...  
194  http://images.amazon.com/images/P/0345313860.0...  
195  http://images.amazon.com/images/P/0316284955.0...  

[196 rows x 7 columns]

In [167]:
famous_books.merge(books, on="book_name").drop_duplicates("book_name")

book_name  total_ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)            428   
3         Harry Potter and the Goblet of Fire (Book 4)            387   
5       Harry Potter and the Sorcerer's Stone (Book 1)            278   
9    Harry Potter and the Order of the Phoenix (Boo...            347   
13    Harry Potter and the Chamber of Secrets (Book 2)            556   
16   The Hobbit : The Enchanting Prelude to The Lor...            281   
17   The Fellowship of the Ring (The Lord of the Ri...            368   
26   Harry Potter and the Sorcerer's Stone (Harry P...            575   
28      The Two Towers (The Lord of the Rings, Part 2)            260   
39                               To Kill a Mockingbird            510   
47                                   The Da Vinci Code            898   
53                  The Five People You Meet in Heaven            430   
55                              The Catcher in the Rye            449   
62                           The Lovely Bones: A Novel           1295   
63                                                1984            284   
72                            Prodigal Summer: A Novel            253   
73                                          Neverwhere            265   
78                             The Secret Life of Bees            774   
84   Stupid White Men ...and Other Sorry Excuses fo...            283   
85   Tuesdays with Morrie: An Old Man, a Young Man,...            493   
88                 The Red Tent (Bestselling Backlist)            723   
89                The Hitchhiker's Guide to the Galaxy            268   
98    Nickel and Dimed: On (Not) Getting By in America            335   
100                                      Into the Wild            252   
103                                     Fahrenheit 451            409   
116                          Girl with a Pearl Earring            526   
117                                          Outlander            283   
123  Where the Heart Is (Oprah's Book Club (Paperba...            585   
124                     Seabiscuit: An American Legend            275   
126                                         Life of Pi            664   
130                                     The Bean Trees            389   
133  A Child Called \It\": One Child's Courage to S...            265   
135                                     ANGELA'S ASHES            279   
137                                        Good in Bed            490   
139  Fast Food Nation: The Dark Side of the All-Ame...            321   
140                                      American Gods            302   
141                                 Skipping Christmas            322   
147  Left Behind: A Novel of the Earth's Last Days ...            318   
149    The Golden Compass (His Dark Materials, Book 1)            336   
157                                   The Color Purple            314   
163                               Silence of the Lambs            256   
166                                        About a Boy            262   
173                  Seven Up (A Stephanie Plum Novel)            278   
175  The Alchemist: A Fable About Following Your Dream            266   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...            269   
182                       Suzanne's Diary for Nicholas            457   
186                                  Lord of the Flies            259   
192                            A Prayer for Owen Meany            413   
194   The Vampire Lestat (Vampire Chronicles, Book II)            301   
195       White Oleander : A Novel (Oprah's Book Club)            356   

     avg_ratings        isbn                  author  \
0       5.852804  0439136350           J. K. Rowling   
3       5.824289  0439139597           J. K. Rowling   
5       5.737410  0590353403           J. K. Rowling   
9       5.501441  043935806X           J. K. Rowling   
13      5.183453  0439064872           J.

In [168]:
famous_books.merge(books, on="book_name").drop_duplicates("book_name").shape

(50, 7)

## Top 50 Books

In [169]:
df = famous_books.merge(books, on="book_name").drop_duplicates("book_name")
df.head()

book_name  total_ratings  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)            428   
3        Harry Potter and the Goblet of Fire (Book 4)            387   
5      Harry Potter and the Sorcerer's Stone (Book 1)            278   
9   Harry Potter and the Order of the Phoenix (Boo...            347   
13   Harry Potter and the Chamber of Secrets (Book 2)            556   

    avg_ratings        isbn         author   publisher  \
0      5.852804  0439136350  J. K. Rowling  Scholastic   
3      5.824289  0439139597  J. K. Rowling  Scholastic   
5      5.737410  0590353403  J. K. Rowling  Scholastic   
9      5.501441  043935806X  J. K. Rowling  Scholastic   
13     5.183453  0439064872  J. K. Rowling  Scholastic   

                                            image_url  
0   http://images.amazon.com/images/P/0439136350.0...  
3   http://images.amazon.com/images/P/0439139597.0...  
5   http://images.amazon.com/images/P/0590353403.0...  
9   http://images.amazon.com/images/P/043935806X.0...  
13  http://images.amazon.com/images/P/0439064872.0...

## Collaborative Filtering